### PR TITLE
Add gauge with timestamp of `trace-ingestion-synthetics`' span end time

### PR DESCRIPTION
**Pull request by plantfansam**
https://github.com/Shopify/service-tracing/issues/443

Working in staging:

```
# on staging pod
curl localhost:8096/metrics

...

# HELP span_metrics_staging_span_metric_ingest_latency
# TYPE span_metrics_staging_span_metric_ingest_latency gauge
span_metrics_staging_span_metric_ingest_latency 1.690841877e+09 1690841879240
```

### CODE REVIEW

File: processor/spanmetricsprocessor/factory.go [hunk 1]
```diff
@@ -47,6 +47,7 @@ func createDefaultConfig() component.Config {
 		DimensionsCacheSize:    defaultDimensionsCacheSize,
 		skipSanitizeLabel:      featuregate.GlobalRegistry().IsEnabled(dropSanitizationGateID),
 		MetricsFlushInterval:   15 * time.Second,
+		IngestLatencyMetricService: "trace-ingestion-synthetics",
```
Comment (rdooley) [inline processor/spanmetricsprocessor/factory.go]: imo dont default this to what appears to be a real service

File: processor/spanmetricsprocessor/processor_test.go [hunk 2]
```diff
@@ -468,7 +475,7 @@ func newProcessorImp(mexp *mocks.MetricsExporter, tcon *mocks.TracesConsumer, de
 	}
 	return &processorImp{
 		logger:          logger,
-		config:          Config{AggregationTemporality: temporality},
+		config:          Config{AggregationTemporality: temporality, IngestLatencyMetricService: "trace-ingestion-synthetics"},
```
Comment (rdooley) [inline processor/spanmetricsprocessor/processor_test.go]: maybe this should be a test value

File: processor/spanmetricsprocessor/config.go [hunk 3]
```diff
@@ -72,6 +72,10 @@ type Config struct {
 
 	// MetricsEmitInterval is the time period between when metrics are flushed or emitted to the configured MetricsExporter.
 	MetricsFlushInterval time.Duration `mapstructure:"metrics_flush_interval"`
+
+	// IngestLatencyMetricService is the service whose spans will generate metrics about ingestion lag in the span metrics
+	// pipeline
+	IngestLatencyMetricService string `mapstructure:"ingest_latency_metrics_service"`
```
Comment (rdooley) [inline processor/spanmetricsprocessor/config.go]:
```suggestion
	IngestLatencyServiceName string `mapstructure:"ingest_latency_service_name"`
```

File: processor/spanmetricsprocessor/processor_test.go (hunk 2 above)
Comment (plantfansam) [reply processor/spanmetricsprocessor/processor_test.go]: Yeah, I felt weird about hard-coding it, to be sure.

File: processor/spanmetricsprocessor/processor.go [hunk 4]
```diff
@@ -371,6 +376,17 @@ func (p *processorImp) collectCallMetrics(ilm pmetric.ScopeMetrics) {
 	}
 }
 
+// collectIngestLatencyMetrics sets up a gauge metric for ingest latency and
+// populates it with the contents of p.ingestLatencyTimestamp.
+func (p *processorImp) collectIngestLatencyMetrics(ilm pmetric.ScopeMetrics) {
```
Comment (rdooley) [inline processor/spanmetricsprocessor/processor.go]:
```suggestion
func (p *processorImp) collectIngestLatency(ilm pmetric.ScopeMetrics) {
```
first off, its one metric not metrics, second allow me to pedantically say metric is implied

File: processor/spanmetricsprocessor/processor.go [hunk 5]
```diff
@@ -412,6 +428,11 @@ func (p *processorImp) aggregateMetrics(traces ptrace.Traces) {
 					asc = uint64(1 << pval)
 				}
 				p.updateHistogram(key, latencyInMilliseconds, span.TraceID(), span.SpanID(), asc)
+
+				if serviceName == p.config.IngestLatencyMetricService {
```
Comment (rdooley) [inline processor/spanmetricsprocessor/processor.go]: Do we need to do this check if the ingest latency service name is not set?

File: processor/spanmetricsprocessor/config.go (hunk 3 above)
Comment (plantfansam) [reply processor/spanmetricsprocessor/config.go]: 👍 5d8f42f361eba57c0f78eb390b1dbec28b291f97

File: processor/spanmetricsprocessor/processor.go (hunk 4 above)
Comment (plantfansam) [reply processor/spanmetricsprocessor/processor.go]: I'm fine changing it, should note that I was trying to stay in line [with package style](https://github.com/Shopify/opentelemetry-collector-contrib/blob/5d8f42f361eba57c0f78eb390b1dbec28b291f97/processor/spanmetricsprocessor/processor.go#L311-L313):

```go
	p.collectIngestLatencyMetrics(ilm)
	p.collectCallMetrics(ilm)
	p.collectLatencyMetrics(ilm)
```

File: processor/spanmetricsprocessor/processor.go (hunk 5 above)
Comment (plantfansam) [reply processor/spanmetricsprocessor/processor.go]: Probably not, but I guess we'd be swapping out a string comparison for a `nil` comparison, right? Not sure what's best, performance-wise.

File: processor/spanmetricsprocessor/processor.go (hunk 4 above)
Comment (rdooley) [reply processor/spanmetricsprocessor/processor.go]: i would say its a latency metric, stuff it in that method and execute if the config opt is set

File: processor/spanmetricsprocessor/processor.go (hunk 5 above)
Comment (rdooley) [reply processor/spanmetricsprocessor/processor.go]: what if we dont configure this, and a span is reported with a service of `""`

Comment (plantfansam) [reply processor/spanmetricsprocessor/processor.go]: Yeah, this is a pickle.

Comment (plantfansam) [reply processor/spanmetricsprocessor/processor.go]: I'm not sure if there's a more conventional way to do this, but 6310efdf79cd5767438bc27f2a2ac6b5446436d2

File: processor/spanmetricsprocessor/processor.go (hunk 4 above)
Comment (plantfansam) [reply processor/spanmetricsprocessor/processor.go]: edfd0226a0d9554c360a3c96c4d1406113184911

File: processor/spanmetricsprocessor/processor_test.go (hunk 2 above)
Comment (plantfansam) [reply processor/spanmetricsprocessor/processor_test.go]: 7128e50149b2f71fbd36453a2a230ad24f18e53a

File: processor/spanmetricsprocessor/processor.go [hunk 6]
```diff
@@ -433,6 +451,11 @@ func (p *processorImp) aggregateMetrics(traces ptrace.Traces) {
 					}
 				}
 				p.updateHistogram(key, latencyInMilliseconds, span.TraceID(), span.SpanID(), asc)
+
+				if p.config.IngestLatencyServiceName != "" && p.config.IngestLatencyServiceName == serviceName {
+					p.updateIngestLatency(span.StartTimestamp())
```
Comment (erin-noe-payne) [inline processor/spanmetricsprocessor/processor.go]: I think I care about the EndTimestamp

**Review by erin-noe-payne — APPROVED** [review]

File: processor/spanmetricsprocessor/processor.go (hunk 6 above)
Comment (plantfansam) [reply processor/spanmetricsprocessor/processor.go]: bd347033f76b792d5a3e2ac3eace8ccfe86bdfb8

**Pull request merged**